routing/http/types/iter/iter.go

```diff
@@ -0,0 +1,28 @@
+package iter
+
+// Iter is an iterator of aribtrary values.
+// Iterators are generally not goroutine-safe, to make them safe just read from them into a channel.
+// For our use cases, these usually have a single reader. This motivates iterators instead of channels,
+// since the overhead of goroutines+channels has a significant performance cost.
+// Using an iterator, you can read results directly without necessarily involving the Go scheduler.
+type Iter[T any] interface {
+	// Next returns the next element, true if an attempt was made to get the next element, and any error that occurred.
+	Next() (T, bool, error)
```

Comment:
WDYT about:

```suggestion
	Next(ctx context.Context) (T, error)
	io.Closer
```
Where instead of using a boolean to know if we still have new elements, we use `io.EOF`? (this just simplify the method definition and also a little bit the code and the implementation IMHO)

Reply:
I don't feel strongly about either. I always found `io.EOF` to be clunky since it's not really an error so it's easy to miss or handle incorrectly due to being mislabeled, but OTOH it's still idiomatic and having three return values can also be confusing.  I'll try the EOF and see how it looks.

Reply:
I also don't think a ctx is necessary (including in JSONIter where I included one), this is controlled by the underlying reader being closed (e.g. when reading client-side and the req context is canceled, the body reader is closed and the context err is returned).

Reply:
I tried it with `io.EOF` and it was pretty annoying to have to do `if errors.Is(err, io.EOF){...}` instead of `if ok {...}`.  Definitely adding `Close()` though.

Reply:
ACK

Reply:
Sorry, after reviewing the code, I changed my mind :rofl: . Implementation-wise, using 2 return parameters instead of 3 has several advantages:
- Less error-prone. The developer doesn't have to decide when the returned boolean has to be true or false, it depends only on the error output. 
- Easy to chain iterators, less code doing the same. Check: https://github.com/ipfs/go-libipfs/pull/18#discussion_r1072012310
- Easy to integrate with existing functions. There is no example on the Go SDK doing what we are doing with the Iterator, returning an error and an ok flag value simultaneously. That makes integration with existing APIS more verbose. See example here: https://github.com/ipfs/go-libipfs/pull/18#discussion_r1072015158

Reply:
> Less error-prone. The developer doesn't have to decide when the returned boolean has to be true or false, it depends only on the error output.

I disagree with this characterization, because you have to remember to check io.EOF. Typically I associate that with io.Reader, not other random non-io fns.

The other points are ultimately about ease of use, which is a secondary concern to me. Iterating over a list of stuff that does preemptible I/O, and could error out at any point, is not easy, and the complex interface reflects that instead of hiding it behind a sentinel error.

Reply:
> I disagree with this characterization, because you have to remember to check io.EOF

This is the same as saying that you have to remember to check the flag... you have to check something to know when to stop iterating. Using a sentinel error makes to only have one place to check. Errors that need to be returned by an interface must be described on the Godocs methods, that's how people will know how to check when the iterator is finished.

My point is that this Iterator API is like IO API, if it returns an error, then you must stop reading, so having a boolean flag duplicates that information. Every time an error is returned on an iterator implementation, the flag will be false.

That's why I think a sentinel error like io.EOF is a good idea here: we simplify the implementations, we simplify the interaction with other standard libraries, and we avoid duplicated method outputs.

Reply:
@guseggert WDYT about https://pkg.go.dev/database/sql#Rows.Scan%5D approach?

```go
if iter.Next() {
    val := iter.Value()
    ....
}
if err := iter.Err(); err != nil {
    panic(err)
}
```

Reply:
sgtm thanks for finding this